pallets/vesting/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+					);
+
+					T::Currency::set_lock(VESTING_LOCK_ID, who, total_amount, WithdrawReasons::all());
+					VestingSchedules::<T>::insert(who, bounded_schedules);
+				});
+		}
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::claim(<T as Config>::MaxVestingSchedules::get() / 2))]
+		pub fn claim(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let locked_amount = Self::do_claim(&who);
+
+			Self::deposit_event(Event::Claimed {
+				who,
+				amount: locked_amount,
+			});
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::vested_transfer())]
+		pub fn vested_transfer(
+			origin: OriginFor<T>,
+			dest: <T::Lookup as StaticLookup>::Source,
+			schedule: VestingScheduleOf<T>,
+		) -> DispatchResult {
+			let from = T::VestedTransferOrigin::ensure_origin(origin)?;
+			let to = T::Lookup::lookup(dest)?;
+
+			if to == from {
+				ensure!(
+					T::Currency::free_balance(&from) >= schedule.total_amount().ok_or(ArithmeticError::Overflow)?,
+					Error::<T>::InsufficientBalanceToLock,
+				);
+			}
+
+			Self::do_vested_transfer(&from, &to, schedule.clone())?;
+
+			Self::deposit_event(Event::VestingScheduleAdded {
+				from,
+				to,
+				vesting_schedule: schedule,
+			});
+			Ok(())
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::update_vesting_schedules(vesting_schedules.len() as u32))]
+		pub fn update_vesting_schedules(
+			origin: OriginFor<T>,
+			who: <T::Lookup as StaticLookup>::Source,
+			vesting_schedules: Vec<VestingScheduleOf<T>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let account = T::Lookup::lookup(who)?;
+			Self::do_update_vesting_schedules(&account, vesting_schedules)?;
+
+			Self::deposit_event(Event::VestingSchedulesUpdated { who: account });
+			Ok(())
+		}
+
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::claim(<T as Config>::MaxVestingSchedules::get() / 2))]
+		pub fn claim_for(origin: OriginFor<T>, dest: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+			let _ = ensure_signed(origin)?;
+			let who = T::Lookup::lookup(dest)?;
+			let locked_amount = Self::do_claim(&who);
+
+			Self::deposit_event(Event::Claimed {
+				who,
+				amount: locked_amount,
+			});
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_claim(who: &T::AccountId) -> BalanceOf<T> {
+		let locked = Self::locked_balance(who);
+		if locked.is_zero() {
+			// cleanup the storage and unlock the fund
+			<VestingSchedules<T>>::remove(who);
+			T::Currency::remove_lock(VESTING_LOCK_ID, who);
+		} else {
+			T::Currency::set_lock(VESTING_LOCK_ID, who, locked, WithdrawReasons::all());
+		}
+		locked
+	}
+
+	/// Returns locked balance based on current block number.
+	fn locked_balance(who: &T::AccountId) -> BalanceOf<T> {
```

Comment:
add unit test

Reply:
Suggestion: Remove function so that it is not reused between `do_claim` and `do_vested_transfer`